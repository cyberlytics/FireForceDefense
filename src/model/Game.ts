import LevelManager from './LevelManager';
import type LevelDefinition from './LevelDefinition';
import LevelMap from './LevelMap';
import Loeschkran from '../contents/Loeschkran';
import Loeschturm from '../contents/Loeschturm';
import Loeschzeppelin from '../contents/Loeschzeppelin';
import Loeschschiff from '../contents/Loeschschiff';
import Loeschtrupp from '../contents/Loeschtrupp';
import type Cell from './Cell';
import type ContentDerivedType from './ContentDerivedType';
import type HexPosition from './HexPosition';
import type ContentDefinition from './ContentDefinition';
import Basis from '../contents/Basis';
import Abgebrannt from '../cells/Abgebrannt';
import type Content from './Content';
import Brandreste from '../contents/Brandreste';
import Fire from './Fire';

export default class Game {
    private levelDefinition: LevelDefinition;
    private readonly levelMap: LevelMap;

    private _contentToBuild: ContentDerivedType = null;
    private _isBaseBuilt = false;

    private get isBaseBuilt() {
        return this._isBaseBuilt;
    }

    private set isBaseBuilt(value) {
        if (!this._isBaseBuilt && value) {
            this.start();
        } else if (this._isBaseBuilt && !value) {
            // This could be used to rebuild the base after it was removed in some non-level-ending way.
            // As of now, no such use case exists.
            // This is mainly implemented as a symmetrical counterpart to the if-statement above.
            this.pause();
            this.contentToBuild = Basis;
        }
        this._isBaseBuilt = value;
    }

    // How long a game step should last in milliseconds
    private gameStepDuration = 1000;

    // Milliseconds since 1970-01-01 at the time the current game step timeout was started
    private gameStepTimeoutStart: number|null = null;

    // Remaining time for the game step timeout
    private gameStepRemainingTime: number|null = null;

    // ID for the game step timeout as returned by setTimeout
    private gameStepTimeoutID: number|null = null;

    // Number of fully executed game steps
    private gameStepCounter = 0;

    get contentToBuild() {
        return this._contentToBuild;
    }

    set contentToBuild(value) {
        if (!this.isBaseBuilt && value !== Basis) {
            return;
        }
        this._contentToBuild = value;
        this.markDisabledCells();
    }

    constructor(levelID: string) {
        const lm = LevelManager.getInstance();
        const ld = lm.getLevelDefinition(levelID);
        if (!ld) {
            throw new Error('Level not registered!');
        }
        this.levelDefinition = ld;
        // TODO fetch user score and check if level is unlocked
        this.levelMap = new LevelMap(this.levelDefinition.cellDefinitions);

        ld.contentDefinitions.forEach((contentDefinition: ContentDefinition) => {
            this.placeContentAt(new contentDefinition.contentType(), contentDefinition.pos);
        });

        if (!this.isBaseBuilt) {
            this.contentToBuild = Basis;
        }
    }

    public static getBuildableContents() {
        return [
            Loeschkran,
            Loeschturm,
            Loeschzeppelin,
            Loeschschiff,
            Loeschtrupp,
        ];
    }

    public getLevelMap() {
        return this.levelMap;
    }

    public placeAt(position: HexPosition): boolean {
        if (this.contentToBuild === null) {
            return false;
        }
        if (this.placeContentAt(new this.contentToBuild(), position)) {
            // TODO Remove build costs from money here
            return true;
        }
        return false;
    }

    private placeContentAt(content: Content, position: HexPosition): boolean {
        const cell = this.levelMap.getCellAt(position);
        if (cell.content !== null) {
            return false;
        }
        if (content.isPlaceableOn(cell)) {
            if (content.id === 'Basis') {
                this.isBaseBuilt = true;
            }
            cell.content = content;
        } else {
            return false;
        }
    }

    private markDisabledCells() {
        let func = (cell: Cell) => false;
        if (this.contentToBuild !== null) {
            func = (cell: Cell) => !new (this.contentToBuild)().isPlaceableOn(cell) || cell.content !== null;
        }
        this._isCellDisabled = func;
    }

    private _isCellDisabled: (cell: Cell) => boolean = () => false;

    public getCellDisabledFunction() {
        return this._isCellDisabled;
    }

    private fireDamage() {
        this.levelMap.getAllCells().forEach((cell) => {
            if (cell.content && cell.content.applyDamage(cell.fireIntensity)) {
                cell.content = new Brandreste();
            }
            if (cell.applyDamage()) {
                const content = cell.content;
                const intensity = cell.fireIntensity;
                const burnt = new Abgebrannt(cell.position);
                this.levelMap.replaceCell(burnt);
                this.levelMap.getCellAt(cell.position).fireIntensity = intensity;
                if (content) {
                    this.placeContentAt(content, cell.position);
                }
            }
        });
    }

    private ownFireChange() {
        this.levelMap.getAllCells().forEach((cell) => {
            cell.fireIntensity = Fire.modify(cell.fireIntensity, cell.fireGrowAmount, cell.maxFireIntensity, true);
        });
    }

    private step() {
        this.fireDamage();

        // TODO Check if the base still exists and end the level if not

        // TODO Calculate fire spread to neighbor cells

        // TODO Apply fire spread to neighbor cells

        // TODO Reduce fire intensity according to the fire extinguishing contents

        this.ownFireChange();

        // TODO Check if there are still fires burning and end the level if not

        // TODO Execute effects
    }

    private timeoutHandler() {
        this.start();
        this.step();
        this.gameStepCounter++;
    }

    public start() {
        this.gameStepTimeoutID = window.setTimeout(
            () => this.timeoutHandler(),
            this.gameStepRemainingTime === null ? this.gameStepDuration : this.gameStepRemainingTime
        );
        this.gameStepTimeoutStart = Date.now();
        this.gameStepRemainingTime = null;
    }

    public pause() {
        window.clearTimeout(this.gameStepTimeoutID);
        this.gameStepRemainingTime = this.gameStepDuration + this.gameStepTimeoutStart - Date.now();
        this.gameStepTimeoutID = null;
        this.gameStepTimeoutStart = null;
    }
}
