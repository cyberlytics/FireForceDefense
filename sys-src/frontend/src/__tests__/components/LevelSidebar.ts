import { mount, RouterLinkStub } from '@vue/test-utils';
import LevelSidebar from '@components/levelSidebar.vue';
import Game from '@model/Game';

describe('LevelSidebar.vue', () => {
    test('level sidebar matches snapshot', () => {
        const wrapper = mount(LevelSidebar, {
            propsData: {
                reliefGotActivated: false,
                buildableContents: Game.getBuildableContents(),
                helpTexts: [],
                totalMoney: 42,
                modalId: 'level-menu-modal',
            },
            mocks: {
                $t: (key: string) => key,
            },
            stubs: {
                RouterLink: RouterLinkStub,
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    test('emit "set help text" when hovering the money display', () => {
        const wrapper = mount(LevelSidebar, {
            propsData: {
                reliefGotActivated: false,
                buildableContents: Game.getBuildableContents(),
                helpTexts: [],
                totalMoney: 42,
                modalId: 'level-menu-modal',
            },
            mocks: {
                $t: (key: string) => key,
            },
            stubs: {
                RouterLink: RouterLinkStub,
            },
        });
        wrapper.find('#money').trigger('mouseenter');
        wrapper.find('#money').trigger('mouseleave');
        expect(typeof wrapper.emitted()['set-help-text'][0][0]).toBe('string');
        expect(wrapper.emitted()['set-help-text'][0][0]).not.toHaveLength(0);
        expect(wrapper.emitted()['set-help-text'][1][0]).toBeNull();
    });

    // TODO more tests can be added later, for example:
    // - Relief activated: true/false
    // - Hovering over all kinds of elements
    // - Setting the help text
});
