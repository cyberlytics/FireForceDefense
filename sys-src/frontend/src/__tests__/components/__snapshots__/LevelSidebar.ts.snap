// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LevelSidebar.vue level sidebar matches snapshot 1`] = `
<div
  id="level-view-sidebar"
  style="z-index: 100;"
>
  <div
    class="d-flex flex-row justify-content-between align-items-center mb-3 row-cols-3"
  >
    <div
      class="d-flex justify-content-start"
    >
      <div
        class="bg-light p-2"
        id="money"
      >
        
                42 
        <img
          alt="Coins"
          src="/assets/utilities-Coin.svg"
          style="height: 1em;"
        />
      </div>
    </div>
     
    <div
      class="d-flex justify-content-center"
    >
      <button
        class="btn btn-card"
        id="rain-emergency-relief"
      >
        <img
          alt="Rain"
          src="/assets/utilities-Nothilfe.svg"
          style="height: 3em;"
        />
      </button>
    </div>
     
    <div
      class="d-flex justify-content-end"
    >
      <button
        class="btn btn-light"
        data-target="#level-menu-modal"
        data-toggle="modal"
      >
        <img
          alt="Menu"
          src="/assets/utilities-menu_bttn.svg"
          style="height: 1em;"
        />
      </button>
    </div>
  </div>
   
  <div
    class="card my-2"
  >
    <div
      class="card-body"
    >
      <div
        class="d-flex justify-content-between mb-2"
      >
        <h2
          class="card-title h5"
        >
          building menu
        </h2>
         
        <button
          class="small-menu-card btn btn-card"
        >
          <svg
            height="100%"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <rect
              fill="url(#utility-Abbauen)"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
        </button>
      </div>
       
      <div
        class="build-menu-cards"
      >
        <button
          class="build-menu-card btn btn-card"
        >
          <svg
            class="p-1"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 100 100"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <rect
              fill="url(#content-Loeschtrupp)"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
           
          <span
            class="d-flex justify-content-end mt-1"
          >
            <span
              class="build-costs"
            >
              10
            </span>
          </span>
        </button>
        <button
          class="build-menu-card btn btn-card"
        >
          <svg
            class="p-1"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 100 100"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <rect
              fill="url(#content-Loeschturm)"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
           
          <span
            class="d-flex justify-content-end mt-1"
          >
            <span
              class="build-costs"
            >
              25
            </span>
          </span>
        </button>
        <button
          class="build-menu-card btn btn-card"
        >
          <svg
            class="p-1"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 100 100"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <rect
              fill="url(#content-Loeschkran)"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
           
          <span
            class="d-flex justify-content-end mt-1"
          >
            <span
              class="build-costs"
            >
              40
            </span>
          </span>
        </button>
        <button
          class="build-menu-card btn btn-card"
        >
          <svg
            class="p-1"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 100 100"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <rect
              fill="url(#content-Loeschschiff)"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
           
          <span
            class="d-flex justify-content-end mt-1"
          >
            <span
              class="build-costs"
            >
              100
            </span>
          </span>
        </button>
        <button
          class="build-menu-card btn btn-card"
        >
          <svg
            class="p-1"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 100 100"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <rect
              fill="url(#content-Loeschzeppelin)"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
           
          <span
            class="d-flex justify-content-end mt-1"
          >
            <span
              class="build-costs"
            >
              150
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
   
  <div
    class="card my-2"
    id="help"
  >
    <div
      class="card-body"
    >
      <h2
        class="card-title h5"
      >
        help
      </h2>
       
    </div>
  </div>
</div>
`;
